pymeasure/instruments/hp/hp8116a.py

```diff
@@ -0,0 +1,506 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2021 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+import time
+import numpy as np
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set, strict_range, truncated_discrete_set
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+class HP8116A(Instrument):
+    """ Represents the Hewlett-Packard 8116A 50 MHz Pulse/Function Generator
+    and provides a high-level interface for interacting with the instrument.
+    The resolution for all floating point instrument parameters is 3 digits.
+    """
+
+    def __init__(self, resourceName, **kwargs):
+        kwargs.setdefault('read_termination', '\r\n')
+        kwargs.setdefault('write_termination', '\r\n')
+        kwargs.setdefault('send_end', True)
+        super(HP8116A, self).__init__(
+            resourceName,
+            'Hewlett-Packard 8116A Pulse/Function Generator',
+            includeSCPI=False,
+            **kwargs
+        )
+    
+    OPERATING_MODES = {
+        'normal': 'M1',
+        'triggered': 'M2',
+        'gate': 'M3',
+        'external_width': 'M4',
+        
+        # Option 001 only
```

Comment:
No, unfortunately there is no direct way like a command. I think I'd have to send a command that's only valid with Option 001 and see if there's an error.
But now that you mentioned it, I definetely have to know wether the connected unit has Option 001, since the status responses have different lengths. However I would say it would be better to have it as a constructor argument, what do you think?

Reply:
Yes, I still have to implement `check_errors`. There is an "IERR" query and also a status byte that contains error information. Also, there are still some other small features that I have to implement like autovernier mode.
Thanks for pointing out the documentation, I'm going to do that when the implementation is ready.

Reply:
@CodingMarco Wanted to propose some enum action for the errors but had a look at the manual of the 8116A, and the `IERR` comand is actually quite comfortable with the clear ASCII replies (e.g. `LIMIT ERROR`, tables 3-6 and 3-7 of the manual).

Reply:
> No, unfortunately there is no direct way like a command. I think I'd have to send a command that's only valid with Option 001 and see if there's an error. But now that you mentioned it, I definetely have to know wether the connected unit has Option 001, since the status responses have different lengths. However I would say it would be better to have it as a constructor argument, what do you think?

Actually there is are two ways to see if option 001 is installed:

- like you mentioned before: Set an operation mode (e.g. M5-M8 which are Option 001 only) and if the option is not installed, we should get an errorstatus.

- length of the CST (current status, Paragraph 3-68 of the manual) return data, without option 001 the length should be 89 bytes, with option 001 its length is 161 bytes. so using this command and parsing it during the initialization of the instrument instrance could give us this info and also fill the instrument properties with the current status of the instrument.